src/common/libsubprocess/ev_fbuf_read.c

```diff
@@ -105,6 +110,31 @@ static void check_cb (struct ev_loop *loop, ev_check *w, int revents)
 
         if (is_eof)
             ebr->eof_sent = true;
+        else if (!fbuf_space (ebr->fb)) {
+            /* At the end of the day, there is a core assumption that
+             * users will not ignore reading data when they are told
+             * there is data to read.
+             *
+             * If the user didn't read anything above and we're out of
+             * buffer space, we gotta do something otherwise we will
+             * spin (i.e. the io watcher is currently stopped, it
+             * can't be restarted b/c the user isn't reading data,
+             * etc.)
+             *
+             * we could stop the ev watchers (prep, check, idle, and
+             * io), but this results in little ability to control
+             * "fallout" from a watcher just (effectively) exiting out
+             * of the blue.  From caller perspectives, it may have
+             * exited cleanly.
+             *
+             * we choose to dump the buffer contents instead.
+             * Unfortunately this leads to loss of data and no error
+             * message.  However in authors opinion, it is a "cleaner"
+             * fallout.
+             *
+             */
```

Comment:
Pondering this a bit, it seems like this ev watcher ought to behave similar to the libev ev_io watcher:

http://pod.tst.eu/http://cvs.schmorp.de/libev/ev.pod#code_ev_io_code_is_this_file_descrip

That is, it should be level triggered (keep calling callback while data is available) and should be stoppable in the watcher callback.  This change would drop data if the watcher chose to stop itself rather than consume the data so I think it would not be correct.

It may be good to add some inline documentation though (I guess in `fbuf_watcher.h`?) that the watchers are level triggered and that the read watcher must either read some data or stop the watcher to avoid unnecessary trips through the event loop without progress.

Also, maybe add documentation in subprocess.h that the output callbacks should either consume data or call `flux_subprocess_stream_stop()`.

Reply:
Maybe this PR should just ensure that, when line buffered and the buffer is full, the callback gets even there isn't a line terminator, and level triggering documentation (and any fixes for in-tree inappropriate uses, if any) could go in a separate PR?

Reply:
> Maybe this PR should just ensure that, when line buffered and the buffer is full, the callback gets called even there isn't a line terminator, and level triggering documentation (and any fixes for in-tree inappropriate uses, if any) could go in a separate PR?

Let me try again to split it up.  I think the reason I didn't for this PR is b/c a user not being able to read a buffer without a line  == a user not reading data b/c they "forgot to".  So that's why I ended up putting it all together.  It led to some testing difficulties otherwise.  But I'll give it another shot.